satpy/dataset/metadata.py

```diff
@@ -137,22 +147,71 @@ def _all_arrays_equal(arrays):
     """
     if hasattr(arrays[0], 'compute'):
         return _all_identical(arrays)
-    else:
-        return _pairwise_all(nan_allclose, arrays)
+    return _all_values_equal(arrays)
+
+
+def _all_values_equal(values):
+    return _all_close(values) or _all_equal(values)
+
+
+def _all_dicts_equal(dicts):
+    return _pairwise_all(_dict_equal, dicts)
+
+
+def _dict_equal(d1, d2):
+    """Check that two dictionaries are equal.
+
+    Nested dictionaries are flattened to facilitate comparison.
+    """
+    d1_flat = flatten_dict(d1)
+    d2_flat = flatten_dict(d2)
+    if not _dict_keys_equal(d1_flat, d2_flat):
+        return False
+    for key in d1_flat.keys():
+        value_pair = [d1_flat[key], d2_flat[key]]
+        if not _all_non_dicts_equal(value_pair):
+            return False
+    return True
+
+
+def _dict_keys_equal(d1, d2):
+    return d1.keys() == d2.keys()
 
 
 def _pairwise_all(func, values):
     for value in values[1:]:
-        if not func(values[0], value):
+        if not _is_equal(values[0], value, func):
             return False
     return True
 
 
+def _is_equal(a, b, comp_func):
+    try:
+        return _array_or_object_equal(a, b, comp_func)
+    except (ValueError, AttributeError, TypeError):
+        return False
+
+
+def _array_or_object_equal(a, b, comp_func):
+    res = comp_func(a, b)
+    if _is_array(res):
```

Comment:
This is needed if `_all_close` fails due to the data type (e.g. `datetime`). Then we go to `_all_equal` and there `a == b` is an array.

Reply:
should this comment be an actual in the code? eg on line 154?